subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c

```diff
@@ -277,30 +278,62 @@ do_firmware_transfer_reply_cb(const struct coap_packet *response,
 	transfer_offset = coap_next_block(check_response, &firmware_block_ctx);
 
 	/* Process incoming data */
-	payload = coap_packet_get_payload_ptr(check_response, &payload_len,
-					      false);
+	payload_frag = coap_packet_get_payload(check_response, &payload_offset,
+					       &payload_len);
 	if (payload_len > 0) {
 		SYS_LOG_DBG("total: %zd, current: %zd",
 			    firmware_block_ctx.total_size,
 			    firmware_block_ctx.current);
 
-		callback = lwm2m_firmware_get_write_cb();
-		if (callback) {
-			ret = callback(0, payload, payload_len,
-				       transfer_offset == 0,
-				       firmware_block_ctx.total_size);
-			if (ret == -ENOMEM) {
-				lwm2m_firmware_set_update_result(
+		/* look up firmware package resource */
+		ret = lwm2m_engine_get_resource("5/0/0", &res);
+		if (ret < 0) {
+			return ret;
+		}
+
+		/* get buffer data */
+		write_buf = res->data_ptr;
+		write_buflen = res->data_len;
+
+		/* check for user override to buffer */
+		if (res->pre_write_cb) {
+			write_buf = res->pre_write_cb(0, &write_buflen);
+		}
+
+		write_cb = lwm2m_firmware_get_write_cb();
+		if (write_cb) {
+			/* flush incoming data to write_cb */
+			while (payload_len > 0) {
+				len = (payload_len > write_buflen) ?
+				       write_buflen : payload_len;
+				payload_len -= len;
+				payload_frag = net_frag_read(payload_frag,
+							     payload_offset,
+							     &payload_offset,
+							     len,
+							     write_buf);
+				/* check for end of packet */
+				if (!payload_frag && payload_offset == 0xffff) {
+					return -EINVAL;
```

Comment:
need to update status before return -EINVAL;
<pre>
lwm2m_firmware_set_update_result(RESULT_INTEGRITY_FAILED)
</pre>

Reply:
how about this one?

Reply:
Woops.  Missed this.  Will repost a new version.